Using alphanumeric sort on parameters of matching precedence

jlstevens opened the pull request:
This PR sorts parameters alphanumerically with the same precedence.

jlstevens committed:
Using alphanumeric sort on parameters of matching precedence

philippjfr commented on paramnb/__init__.py:
```diff
@@ -263,7 +263,12 @@ def widgets(self):
+        sorted_precedence = sorted(params().items(),
+                        key=lambda x: x[1].precedence if x[1].precedence else -float('inf'))
+        groups = itertools.groupby(sorted_precedence, key=lambda p: p[1].precedence
+                                   if p[1].precedence else -float('inf'))
```
Might be an idea to factor the key function out since it's repeated.

jlstevens replied:
Good point. I'll pull it out now.

jbednar replied:
This should also be omitting parameters with a negative precedence, e.g. by slicing the list after it's sorted.  

Once negative precedences are omitted, then there's no particular reason to use float('inf') for precedence of None.  Maybe None should actually have a precedence of something intermediate, like 10, so that with the usual 0 to 1.0 precedence values, anything with no precedence goes after those with declared precedence, while still allowing people to force something to appear after all those with no declared precedence (by setting a precedence > 10)?  10 seems arbitrary, but there wouldn't seem to be any particular boundary that would make more sense than that...

jlstevens committed:
Pulled out key_fn for sorted and groupby

jbednar commented:
(Copied here because the latest commit hides my comment)

This should also be omitting parameters with a negative precedence, e.g. by slicing the list after it's sorted.

Once negative precedences are omitted, then there's no particular reason to use float('inf') for precedence of None. Maybe None should actually have a precedence of something intermediate, like 10, so that with the usual 0 to 1.0 precedence values, anything with no precedence goes after those with declared precedence, while still allowing people to force something to appear after all those with no declared precedence (by setting a precedence > 10)? 10 seems arbitrary, but there wouldn't seem to be any particular boundary that would make more sense than that...

jlstevens commented:
> This should also be omitting parameters with a negative precedence, e.g. by slicing the list after it's sorted.

Sure! That is a new feature that is orthogonal to what I was trying to implement here but I am happy to include that in this PR.

Personally, I would prefer a precedence of None to mean 'zero', falling in line with normal Python semantics.

philippjfr commented:
I'd also prefer if None was equivalent to zero, an arbitrary intermediate value is just going to lead to confusion imo.

jlstevens committed:
Fixed precedence sorting which now ignores negative precedence values

jlstevens commented:
I fixed quite a number of bugs after realizing my 'test' was not using the right copy of paramnb and had gotten the order right by chance!

In the last commit, I made the necessary fixes and implemented the suggestion to ignore negative precedence parameters completely.

philippjfr commented:
Looks good.
